File: go/object/object.go
package object

import (
	"bytes"
	"fmt"
	"hash/fnv"
	"strings"

	"github.com/brattonross/interpreter/ast"
)

type ObjectType string

const (
	INTEGER_OBJ      = "INTEGER"
	BOOLEAN_OBJ      = "BOOLEAN"
	NULL_OBJ         = "NULL"
	RETURN_VALUE_OBJ = "RETURN_VALUE"
	ERROR_OBJ        = "ERROR"
	FUNCTION_OBJ     = "FUNCTION"
	STRING_OBJ       = "STRING"
	BUILTIN_OBJ      = "BUILTIN"
	ARRAY_OBJ        = "ARRAY"
	HASH_OBJ         = "HASH"
)

type Object interface {
	Type() ObjectType
	Inspect() string
}

type Integer struct {
	Value int64
}

func (i *Integer) Inspect() string {
	return fmt.Sprintf("%d", i.Value)
}

func (i *Integer) Type() ObjectType {
	return INTEGER_OBJ
}

func (i *Integer) HashKey() HashKey {
	return HashKey{
		Type:  i.Type(),
		Value: uint64(i.Value),
	}
}

type Boolean struct {
	Value bool
}

func (b *Boolean) Inspect() string {
	return fmt.Sprintf("%t", b.Value)
}

func (b *Boolean) Type() ObjectType {
	return BOOLEAN_OBJ
}

func (b *Boolean) HashKey() HashKey {
	var value int64
	if b.Value {
		value = 1
	} else {
		value = 0
	}

	return HashKey{
		Type:  b.Type(),
		Value: uint64(value),
	}
}

type Null struct{}

func (n *Null) Inspect() string {
	return "null"
}

func (n *Null) Type() ObjectType {
	return NULL_OBJ
}

type ReturnValue struct {
	Value Object
}

func (rv *ReturnValue) Inspect() string {
	return rv.Value.Inspect()
}

func (rv *ReturnValue) Type() ObjectType {
	return RETURN_VALUE_OBJ
}

type Error struct {
	Message string
}

func (e *Error) Inspect() string {
	return "ERROR: " + e.Message
}

func (e *Error) Type() ObjectType {
	return ERROR_OBJ
}

type Function struct {
	Parameters []*ast.Identifier
	Body       *ast.BlockStatement
	Env        *Environment
}

func (f *Function) Inspect() string {
	var out bytes.Buffer

	params := []string{}
	for _, p := range f.Parameters {
		params = append(params, p.String())
	}

	out.WriteString("fn")
	out.WriteString("(")
	out.WriteString(strings.Join(params, ", "))
	out.WriteString(") {\n")
	out.WriteString(f.Body.String())
	out.WriteString("\n}")

	return out.String()
}

func (f *Function) Type() ObjectType {
	return FUNCTION_OBJ
}

type String struct {
	Value string
}

func (s *String) Inspect() string {
	return s.Value
}

func (s *String) Type() ObjectType {
	return STRING_OBJ
}

func (s *String) HashKey() HashKey {
	h := fnv.New64a()
	h.Write([]byte(s.Value))
	return HashKey{
		Type:  s.Type(),
		Value: h.Sum64(),
	}
}

type BuiltinFunction func(args ...Object) Object

type Builtin struct {
	Fn BuiltinFunction
}

func (b *Builtin) Inspect() string {
	return "builtin function"
}

func (b *Builtin) Type() ObjectType {
	return BUILTIN_OBJ
}

type Array struct {
	Elements []Object
}

func (a *Array) Inspect() string {
	var out bytes.Buffer

	elements := []string{}
	for _, el := range a.Elements {
		elements = append(elements, el.Inspect())
	}

	out.WriteString("[")
	out.WriteString(strings.Join(elements, ", "))
	out.WriteString("]")

	return out.String()
}

func (a *Array) Type() ObjectType {
	return ARRAY_OBJ
}

type HashKey struct {
	Type  ObjectType
	Value uint64
}

type HashPair struct {
	Key   Object
	Value Object
}

type Hash struct {
	Pairs map[HashKey]HashPair
}

func (h *Hash) Inspect() string {
	var out bytes.Buffer

	pairs := []string{}
	for _, pair := range h.Pairs {
		pairs = append(pairs, fmt.Sprintf("%s: %s", pair.Key.Inspect(), pair.Value.Inspect()))
	}

	out.WriteString("{")
	out.WriteString(strings.Join(pairs, ", "))
	out.WriteString("}")

	return out.String()
}

func (h *Hash) Type() ObjectType {
	return HASH_OBJ
}

type Hashable interface {
	HashKey() HashKey
}

File: object.ts
import type { BlockStatement, IdentifierExpression } from "./ast";

export type Object = {
  readonly type: string;
  inspect(): string;
};

export type Hashable = {
  hashCode(): number;
};

export class IntegerObject implements Object, Hashable {
  public readonly type = "INTEGER";

  public constructor(public value: number) {}

  public inspect(): string {
    return this.value.toString();
  }

  public hashCode(): number {
    return this.value;
  }
}

export class StringObject implements Object, Hashable {
  public readonly type = "STRING";

  public constructor(public value: string) {}

  public inspect(): string {
    return this.value;
  }

  public hashCode(): number {
    let hash = 0;

    for (let i = 0; i < this.value.length; i++) {
      hash = (hash << 5) - hash + this.value.charCodeAt(i);
      hash = hash & hash;
    }

    return hash;
  }
}

export class BooleanObject implements Object, Hashable {
  public readonly type = "BOOLEAN";

  public constructor(public value: boolean) {}

  public inspect(): string {
    return this.value.toString();
  }

  public hashCode(): number {
    return this.value ? 1 : 0;
  }
}

export class NullObject implements Object {
  public readonly type = "NULL";

  public inspect(): string {
    return "null";
  }
}

export class ReturnValueObject implements Object {
  public readonly type = "RETURN_VALUE";

  public constructor(public value: ObjectType) {}

  public inspect(): string {
    return this.value.inspect();
  }
}

export class ErrorObject implements Object {
  public readonly type = "ERROR";

  public constructor(public message: string) {}

  public inspect(): string {
    return `ERROR: ${this.message}`;
  }
}

export class FunctionObject implements Object {
  public readonly type = "FUNCTION";

  public constructor(
    public parameters: Array<IdentifierExpression>,
    public body: BlockStatement | null,
    public env: Environment
  ) {}

  public inspect(): string {
    let out = "fn(";
    for (let i = 0; i < this.parameters.length; i++) {
      out += this.parameters[i]!.toString();
      if (i !== this.parameters.length - 1) {
        out += ", ";
      }
    }
    out += ") {\n";
    if (this.body) {
      out += this.body.toString();
    }
    out += "\n}";
    return out;
  }
}

export class BuiltinFunctionObject implements Object {
  public readonly type = "BUILTIN_FUNCTION";

  public constructor(public fn: (...args: Array<ObjectType>) => ObjectType) {}

  public inspect(): string {
    return "builtin function";
  }
}

export class ArrayObject implements Object {
  public readonly type = "ARRAY";

  public constructor(public elements: Array<ObjectType>) {}

  public inspect(): string {
    let out = "[";
    for (let i = 0; i < this.elements.length; i++) {
      out += this.elements[i]!.inspect();
      if (i !== this.elements.length - 1) {
        out += ", ";
      }
    }
    out += "]";
    return out;
  }
}

export type HashPair = {
  key: ObjectType;
  value: ObjectType;
};

export class HashObject implements Object {
  public readonly type = "HASH";

  public constructor(public pairs: Map<number, HashPair>) {}

  public inspect(): string {
    let out = "{";
    const keys = Array.from(this.pairs.keys());
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]!;
      const pair = this.pairs.get(key)!;
      out += pair.key.inspect();
      out += ": ";
      out += pair.value.inspect();
      if (i !== keys.length - 1) {
        out += ", ";
      }
    }
    out += "}";
    return out;
  }
}

export type ObjectType =
  | IntegerObject
  | StringObject
  | BooleanObject
  | NullObject
  | ReturnValueObject
  | ErrorObject
  | FunctionObject
  | BuiltinFunctionObject
  | ArrayObject
  | HashObject;

export class Environment {
  #store = new Map<string, ObjectType>();
  #outer: Environment | null = null;

  public static enclosed(outer: Environment): Environment {
    const env = new Environment();
    env.#outer = outer;
    return env;
  }

  public get(name: string): ObjectType | null {
    const obj = this.#store.get(name) ?? null;
    if (obj === null && this.#outer !== null) {
      return this.#outer.get(name);
    }
    return obj;
  }

  public set(name: string, value: ObjectType): ObjectType {
    this.#store.set(name, value);
    return value;
  }
}
